# Project description
The focus of this project is to creating and managing python virtual environments, developing a web application, and deploying it to a cloud service that will make it accessible to the public.

In this project, the dataset that was used is about car sales. The data was analyzed by using the pandas library with the help of plotly.express.

## VERSION 2.0
Fixed all of the issues mentioned in the review except the histogram one. My charts were created with plotly, the histograms were created following plotly's documentation. Those are sorted histograms and not bar charts. (Unless plotly is showing how to make bar charts in their documentation under histograms.)

In [1]:
import pandas as pd
import plotly.express as px
import numpy as np

df = pd.read_csv("/Programming/EDAProject/EDAProject/vehicles_us.csv")


df.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [2]:
df.info()
#We havesome datatypes that need to be converted.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
price           51525 non-null int64
model_year      47906 non-null float64
model           51525 non-null object
condition       51525 non-null object
cylinders       46265 non-null float64
fuel            51525 non-null object
odometer        43633 non-null float64
transmission    51525 non-null object
type            51525 non-null object
paint_color     42258 non-null object
is_4wd          25572 non-null float64
date_posted     51525 non-null object
days_listed     51525 non-null int64
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [3]:
df.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [4]:
df.isna().sum()


price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

In [5]:
df.groupby('is_4wd')['is_4wd'].count()

# Wanted to see the values in the column

is_4wd
1.0    25572
Name: is_4wd, dtype: int64

In [6]:
df['is_4wd'] = df['is_4wd'].fillna(0).astype(bool)
df.groupby('is_4wd')['is_4wd'].count()

# First I filled the NaN values with 0 so I can calculate with the column later on. Changed the type of the column to int from float.

is_4wd
False    25953
True     25572
Name: is_4wd, dtype: int64

In [7]:
df["model_year"] = df["model_year"].fillna(df.groupby('model')["model_year"].transform('median'))
df["cylinders"] = df["cylinders"].fillna(df.groupby('model')["cylinders"].transform('median'))
df["odometer"] = df["odometer"].fillna(df.groupby('model_year')["odometer"].transform('median'))

# Replaced the missing values in the numerical colomuns with the median values and changed the datatypes to int.
df.isna().sum()

price              0
model_year         0
model              0
condition          0
cylinders          0
fuel               0
odometer           1
transmission       0
type               0
paint_color     9267
is_4wd             0
date_posted        0
days_listed        0
dtype: int64

In [8]:
df['odometer'] = df['odometer'].fillna(0)
# found a row where the odometer was 0, probably a new car, this is the fix I applied. Since it is one row of data, there is no significance to it. 

types_to_replace = ['model_year', 'cylinders', 'odometer']
for column in types_to_replace:
    df[column] = df[column].apply(np.int64)

df.isna().sum()    

price              0
model_year         0
model              0
condition          0
cylinders          0
fuel               0
odometer           0
transmission       0
type               0
paint_color     9267
is_4wd             0
date_posted        0
days_listed        0
dtype: int64

In [9]:
df['paint_color'] = df['paint_color'].fillna("unknown")
# paint_color was the only non numberic column with mising values, replaced those values with "unknown".

In [10]:
df['date_posted'] = df['date_posted'].astype('datetime64[ns]')
# Converted the date_posted column to datetime type.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
price           51525 non-null int64
model_year      51525 non-null int64
model           51525 non-null object
condition       51525 non-null object
cylinders       51525 non-null int64
fuel            51525 non-null object
odometer        51525 non-null int64
transmission    51525 non-null object
type            51525 non-null object
paint_color     51525 non-null object
is_4wd          51525 non-null bool
date_posted     51525 non-null datetime64[ns]
days_listed     51525 non-null int64
dtypes: bool(1), datetime64[ns](1), int64(5), object(6)
memory usage: 4.8+ MB


In [12]:
df.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

In [13]:
df.duplicated().sum()
# No duplicates in the dataset.

0

In [14]:
df['manufacturer'] = df['model'].apply(lambda x: x.split()[0])
df.head()

# I think it is a good idea to create a new column with the manufacturer only.

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400        2011          bmw x5       good          6  gas    145000   
1  25500        2011      ford f-150       good          6  gas     88705   
2   5500        2013  hyundai sonata   like new          4  gas    110000   
3   1500        2003      ford f-150       fair          8  gas    161397   
4  14900        2017    chrysler 200  excellent          4  gas     80903   

  transmission    type paint_color  is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown    True  2018-06-23           19   
1    automatic  pickup       white    True  2018-10-19           50   
2    automatic   sedan         red   False  2019-02-07           79   
3    automatic  pickup     unknown   False  2019-03-22            9   
4    automatic   sedan       black   False  2019-04-02           28   

  manufacturer  
0          bmw  
1         ford  
2      hyundai  
3         ford  
4     chrysler

In [20]:
fig = px.histogram(df, x="model_year", nbins=25, histnorm='density',                     
                    color_discrete_sequence=["blue"],
                    barmode='overlay',
                    title='Distribution of Cars by Model Year')
fig.update_layout(xaxis={'categoryorder': 'total descending'})
fig.update_traces(opacity=0.7)
fig.show()

# Distribution of Cars by Condition can provide good information about the market.

In [16]:
fig = px.histogram(df, x="manufacturer",
                   histnorm='density', 
                   color_discrete_sequence=["blue"],
                   barmode='overlay',
                   title='Distribution of Cars by Manufacturer')
fig.update_layout(xaxis={'categoryorder': 'total descending'})
fig.update_traces(opacity=0.7)
fig.show()

# Distribution of Cars by Manufacturer shows the amount of ads per manufacturer.

In [17]:


df_avg_price = df.groupby(['manufacturer', 'model_year'])['price'].mean().reset_index()

colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf', '#20B2AA', '#7B68EE', '#FFA07A']

fig = px.scatter(df_avg_price, x='model_year', y='price', color='manufacturer',
             title='Average Price by Manufacturer and Year',
             color_discrete_sequence=colors,
             labels={'year': 'Year', 'price': 'Average Price', 'manufacturer': 'Manufacturer'})
fig.show()

# Average Price by Manufacturer and Year scatter plot gives a solid idea about it there is a relation between car manufacturer, the age of the  car and the average price.

In [18]:
fig = px.scatter(df, x='model_year', y='days_listed',
             title='Year vs. days listed',
             labels={'model_year': 'Year', 'days_listed': 'Days Listed'})
fig.show()


# Year vs. days listed chart showing us that the oldsmobile and new cars are on the market less.

# FINAL CONCLUSION

The goal of this data research is to find out which factors influence the price of a vehicle. 

The dataframe contains data of 51525 advertisements.

The dataset contained missing values in several columns, 5 to be exact.

The missing values in the model_year column were replaced based on the median values of the specific model of the car from the model column.

The missing values in the cylinders column were replaced based on the median values of the same cylinders column grouped by the model.

Paint_color column's missing values were replaced with 'unknown' since these values are in fact unknown.

Is_4wd column is an interesting one. The datatype is set to numerical values so Python thought the 0s are missing values, these has been replaced with actual 0s and the datatype corrected to boolean.

In the odometer column I have replaced the missing values with the median values based on the model_year column since there is a direct correlation between the car's age and the mileage.

The dataset has no duplicate values.

Created 2 histograms that are organized by descending order used by the plotly.express library. Also created two scatter plots using the same library.

The overall conclusion is this short analysis is that the most ads are from Ford, oldsmobiles are selling fast and their price is getting higher. Newer cars are most expensive than and older car but not oldsmobile car yet.